bootloader/lkenv/lkenv.go

```diff
@@ -0,0 +1,355 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package lkenv
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"hash/crc32"
+	"os"
+
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+)
+
+const SNAP_BOOTSELECT_VERSION = 0x00010001
+
+// const SNAP_BOOTSELECT_SIGNATURE ('S' | ('B' << 8) | ('s' << 16) | ('e' << 24))
+const SNAP_NAME_MAX_LEN = 256
+const SNAP_MODE_LENGTH = 8
+
+/* number of available boot partitions */
+const SNAP_BOOTIMG_PART_NUM = 2
+
+/* Default boot image file name to be used from kernel snap */
+const BOOTIMG_DEFAULT_NAME = "boot.img"
+
+// for accessing the 	Bootimg_matrix
+const (
+	MATRIX_ROW_PARTITION = 0
+	MATRIX_ROW_KERNEL    = 1
+)
+
+type SnapBootSelect struct {
+	/* Contains value BOOTSELECT_SIGNATURE defined above */
+	Signature uint32
+	/* snappy boot select version */
+	Version uint32
+
+	/* snap_mode, one of: 'empty', "try", "trying" */
+	Snap_mode [SNAP_MODE_LENGTH]byte
+	/* current core snap revision */
+	Snap_core [SNAP_NAME_MAX_LEN]byte
+	/* try core snap revision */
+	Snap_try_core [SNAP_NAME_MAX_LEN]byte
+	/* current kernel snap revision */
+	Snap_kernel [SNAP_NAME_MAX_LEN]byte
+	/* current kernel snap revision */
+	Snap_try_kernel [SNAP_NAME_MAX_LEN]byte
+
+	/* GADGET assets: current gadget assets revision */
+	Snap_gadget [SNAP_NAME_MAX_LEN]byte
+	/* GADGET assets: try gadget assets revision */
+	Snap_try_gadget [SNAP_NAME_MAX_LEN]byte
+
+	/**
+	Reboot reason
+	Optional parameter to signal bootloader alternative reboot reasons
+	e.g. recovery/factory-reset/boot asset update
+	*/
+	Reboot_reason [SNAP_NAME_MAX_LEN]byte
+
+	/**
+	 Matrix for mapping of boot img partion to installed kernel snap revision
+	 At image build time:
+			 - snap prepare populates boot args as for other bootloaders
+			 - first column with bootimage part names can be filled at gadget builds
+			   or in the future by snap prepare image time
+			 - ExtractKernelAssets fills mapping for initial boot image
+	 snapd:
+			 - when new kernel snap is installed, ExtractKernelAssets updates mapping
+			   in matrix for bootloader to pick correct kernel snap to use for boot
+			 - snap_mode, snap_try_kernel, snap_try_core behaves same way as with u-boot
+			 - boot partition labels are never modified by snapd at run time
+	 bootloader:
+			 - Finds boot partition to use based on info in matrix and snap_kernel / snap_try_kernel
+			 - bootloaer does not alter matrix, only alters snap_mode
+
+	 [ <bootimg 1 part label> ] [ <currently installed kernel snap revison> ]
+	 [ <bootimg 2 part label> ] [ <currently installed kernel snap revision> ]
+	*/
```

Comment:
I find this diagram slightly confusing,  we don't set the same current kernel in both row?  one has the current and one the previous? or am I misunderstanding this completely ?